src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormattingPassBase.cs

```diff
@@ -175,22 +175,29 @@ protected async Task<List<TextChange>> AdjustIndentationAsync(FormattingContext
 
                     var index = Array.BinarySearch(sourceMappingIndentationScopes, lineStart);
 
-                    if (index < 0 && context.SourceText[lineStart] == '@')
-                    {
-                        // Sometimes we are only off by one in finding a source mapping, for example with a simple if statement:
-                        //
-                        // @|if (true)
-                        //
-                        // The sourceMappingIndentationScopes knows about where the pipe is (ie, after the "@") but we're asking
-                        // for indentation at the line start. In these cases we are better off using the real indentation scope,
-                        // than hoping the one before it is correct.
-                        index = Array.BinarySearch(sourceMappingIndentationScopes, lineStart + 1);
-                    }
-
                     if (index < 0)
                     {
                         // Couldn't find the exact value. Find the index of the element to the left of the searched value.
                         index = (~index) - 1;
+
+                        // If that index comes from within a directive that does not contain the line we are working with,
+                        // we don't want to use its indentation. In the following example, we would not want the indentation
+                        // that applies to line 2 to also be applied to line 4:
+                        //
+                        // 1  @section Foo {
+                        // 2  <div></div>
+                        // 3  }
+                        // 4  <p></p>
+                        for (; index >= 0; index--)
+                        {
+                            var absoluteIndex = sourceMappingIndentationScopes[index];
+                            var scopeOwner = context.CodeDocument.GetSyntaxTree().Root.LocateOwner(new SourceChange(absoluteIndex, 0, string.Empty));
+                            var containingDirective = scopeOwner.FirstAncestorOrSelf<RazorDirectiveSyntax>();
```

Comment:
I think this could be narrowed to check for `@section` specifically, as that is the only direction (I'm pretty sure?) that will exhibit this issue, due to how the code is generated. Which, if I'm right, also means that we can probably get rid of the `FirstAncestorOrSelf` which would be a big perf win. `LocateOwner` should be deterministic, so we can check parents etc. to see where we are.

I think that will be especially important because right now, this will look the whole way up the tree, for every iteration.

Reply:
today, the located owner in such a case is an `UnclassifiedTextLiteral`, with the following lineage:

`UnclassifiedTextLiteral > CSharpCodeBlock > RazorDirectiveBody > RazorDirective`

so by the time we are able to check if it is from a section, we already have the knowledge that it is from a directive. i could change this from `scopeOwner.FirstAncestorOrSelf<RazorDirectiveSyntax>()` to `scopeOwner.Parent?.Parent?.Parent as RazorDirectiveSyntax` instead?